### PR TITLE
[v3]  Add validator in render v2.

### PR DESCRIPTION
**Pull request by yuwenma**
Fixes: #nnn 
**Related**: #5673 
**Merge after**: #5940 

**Description**
- Add the render validator and write the skaffold validation rules to Kptfile pipeline.
- Fixed corner cases when `render.generate` is not given in skaffold.yaml
- More test coverage on Render configs.

To reviewers: 
This PR is diff-based from 5940. Please only review [the last commit](https://github.com/GoogleContainerTools/skaffold/commit/4c4890263ef430ee48538ef370c7cc03f8e61b70)

### CODE REVIEW

**Review by tejal29 — APPROVED** [review]

File: pkg/skaffold/render/validate/validate.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+package validate
+
+import (
+	"fmt"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/render/kptfile"
+	latestV2 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest/v2"
+)
+
+var ValidatorWhitelist = map[string]kptfile.Function{
```
Comment (tejal29) [inline pkg/skaffold/render/validate/validate.go:25]: we could make this private?
```suggestion
var validatorWhitelist = map[string]kptfile.Function{
```

File: pkg/skaffold/render/kptfile/pipeline_type.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	KptFileName       = "Kptfile"
+	KptFileKind       = "Kptfile"
+	KptFileGroup      = "kpt.dev"
+	KptFileVersion    = "v1alpha2"
+	KptFileAPIVersion = KptFileGroup + "/" + KptFileVersion
```
Comment (tejal29) [inline pkg/skaffold/render/kptfile/pipeline_type.go]: consider making this constants private.

File: pkg/skaffold/render/renderer/renderer.go [hunk 3]
```diff
@@ -19,56 +19,83 @@ import (
...
-	generator := generate.NewGenerator(workingDir, *config.Generate)
-	return &SkaffoldRenderer{Generator: *generator, workingDir: workingDir}
+	// TODO(yuwenma): The current work directory may not be accurate if users use --filepath flag.
+	hydrationDir := filepath.Join(workingDir, DefaultHydrationDir)
+
+	var generator *generate.Generator
+	if config.Generate == nil {
+		// If render.generate is not given, default to current working directory.
+		defaultManifests := filepath.Join(workingDir, "*.yaml")
+		generator = generate.NewGenerator(workingDir, latestV2.Generate{Manifests: []string{defaultManifests}})
```
Comment (tejal29) [inline pkg/skaffold/render/renderer/renderer.go:54]: nice!

File: pkg/skaffold/render/renderer/renderer.go [hunk 4]
```diff
@@ -19,56 +19,83 @@ import (
...
+
+	var validator *validate.Validator
+	if config.Validate != nil {
+		var err error
+		validator, err = validate.NewValidator(*config.Validate)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		validator, _ = validate.NewValidator([]latestV2.Validator{})
```
Comment (tejal29) [inline pkg/skaffold/render/renderer/renderer.go:67]: What does this validator do? is it a nil validator?

File: pkg/skaffold/render/renderer/renderer.go [hunk 5]
```diff
@@ -19,56 +19,83 @@ import (
...
-	// TODO(yuwenma): The current work directory may not be accurate if users use --filepath flag.
-	outputPath := filepath.Join(r.workingDir, DefaultHydrationDir)
-	logrus.Debugf("creating render directory: %v", outputPath)
-
-	if err := os.MkdirAll(outputPath, os.ModePerm); err != nil {
-		return fmt.Errorf("creating cache directory for hydration: %w", err)
+	if _, err := os.Stat(r.hydrationDir); os.IsNotExist(err) {
+		logrus.Debugf("creating render directory: %v", r.hydrationDir)
+		if err := os.MkdirAll(r.hydrationDir, os.ModePerm); err != nil {
+			return fmt.Errorf("creating cache directory for hydration: %w", err)
```
Comment (tejal29) [inline pkg/skaffold/render/renderer/renderer.go]:
```suggestion
			return fmt.Errorf("creating render directory for hydration: %w", err)
```

Comment (tejal29) [reply pkg/skaffold/render/renderer/renderer.go]: Is this a fatal error? if its just a cache dir, we can still continue without caching?

File: pkg/skaffold/render/renderer/renderer.go [hunk 6]
```diff
@@ -19,56 +19,83 @@ import (
...
+		}
 	}
-	if _, err := os.Stat(filepath.Join(outputPath, KptFileName)); os.IsNotExist(err) {
-		cmd := exec.CommandContext(ctx, "kpt", "pkg", "init", outputPath)
+	kptFilePath := filepath.Join(r.hydrationDir, kptfile.KptFileName)
+	if _, err := os.Stat(kptFilePath); os.IsNotExist(err) {
+		cmd := exec.CommandContext(ctx, "kpt", "pkg", "init", r.hydrationDir)
 		if _, err := util.RunCmdOut(cmd); err != nil {
 			return err
```
Comment (tejal29) [inline pkg/skaffold/render/renderer/renderer.go]: Is this a kpt user error ? Can we define function and error codes like https://github.com/GoogleContainerTools/skaffold/blob/ca4886fcfbca61a8904190e25328b6ce9abdbd84/pkg/skaffold/deploy/kustomize/errors.go#L24 to define a kpt user error

File: pkg/skaffold/render/renderer/renderer.go [hunk 7]
```diff
@@ -92,11 +119,36 @@ func (r *SkaffoldRenderer) Render(ctx context.Context, out io.Writer, builds []g
...
+	// Read the existing Kptfile content. Kptfile is guaranteed to be exist in prepareHydrationDir.
+	kptFilePath := filepath.Join(r.hydrationDir, kptfile.KptFileName)
+	file, err := os.Open(kptFilePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	kfConfig := &kptfile.KptFile{}
+	if err := yaml.NewDecoder(file).Decode(&kfConfig); err != nil {
+		return fmt.Errorf("unable to parse %v\n, please check if the kptfile is updated to new apiVersion > v1alpha2",
```
Comment (tejal29) [inline pkg/skaffold/render/renderer/renderer.go:128]: thanks for creating an actionable error. Can you use sErrors.New() and add error code as well as suggestion code to this error. 
Docs here
https://github.com/GoogleContainerTools/skaffold/blob/master/DEVELOPMENT.md#adding-actionable-error-messages-to-code

The advantage of this this is, IDEs can consume the suggestion code and decide to show an actionable error message to the user as is or relevant one.

**Review by tejal29 — COMMENTED** [review]
general LGTM.
Few qeustions

File: pkg/skaffold/render/renderer/renderer.go (hunk 4 above)
Comment (yuwenma) [reply pkg/skaffold/render/renderer/renderer.go:67]: Init the validator whose validate functions will be updated if the filewatcher detects a change in the skaffold.yaml.

File: pkg/skaffold/render/renderer/renderer.go (hunk 5 above)
Comment (yuwenma) [reply pkg/skaffold/render/renderer/renderer.go]: since kpt only supports in-place hydration, we do need a tmp dir to hold the DRY configs so as the hydrated config won't pollute the user working dir.

File: pkg/skaffold/render/validate/validate.go (hunk 1 above)
Comment (yuwenma) [reply pkg/skaffold/render/validate/validate.go:25]: done.

File: pkg/skaffold/render/renderer/renderer.go (hunk 6 above)
Comment (yuwenma) [reply pkg/skaffold/render/renderer/renderer.go]: can't guarantee it is a user error or not. I updated the err message with user actions.

File: pkg/skaffold/render/renderer/renderer.go (hunk 7 above)
Comment (yuwenma) [reply pkg/skaffold/render/renderer/renderer.go:128]: oh, we do not want to add the err. If the Decode fails, it basically means the yaml is invalid, and no more details will be given from the Decode return error. That's why we add specific user actions to help users triage.

**Comment by yuwenma** [comment]
Updated. Please take another look @tejal29

File: pkg/skaffold/render/renderer/renderer.go (hunk 7 above)
Comment (tejal29) [reply pkg/skaffold/render/renderer/renderer.go]: we still need an error code with an actionable error message `KPT_INVALID_YAML` or `KPT_USER_ERROR` so we can track these in skaffold and get metrics on it. 
Right, if a skaffold dev fails, the error code logged via metrics pipeline  will be `DEPLOY_UNKNOWN`. We are trying to reduce these overall.

File: pkg/skaffold/render/renderer/renderer.go (hunk 6 above)
Comment (tejal29) [reply pkg/skaffold/render/renderer/renderer.go]: Is there a way to identify all possible if this was a `kpt pkg init` error or user config error?
Does kpt return unique exit codes based on different errors?
We are trying to attach an error code for all cases.  You can also define `KPT_PKG_INIT` as an error code and return here.

File: pkg/skaffold/render/renderer/renderer.go (hunk 7 above)
Comment (yuwenma) [reply pkg/skaffold/render/renderer/renderer.go]: I see. I added some TODO comments (here and other errors in ./render). Will add the Suggestion and error code in the following CL.

File: pkg/skaffold/render/renderer/renderer.go (hunk 6 above)
Comment (yuwenma) [reply pkg/skaffold/render/renderer/renderer.go]: that's a good point. I'm not sure if kpt returns the different exit codes or not. We may need more user data to know how likely a user config error may happen in `kpt pkg init` or `kpt fn`

File: pkg/skaffold/render/renderer/renderer.go (hunk 7 above)
Comment (yuwenma) [reply pkg/skaffold/render/renderer/renderer.go]: @tejal29 Here's the proto PR https://github.com/GoogleContainerTools/skaffold/pull/5953

**Review by tejal29 — APPROVED** [review]
